### ISSUE TITLE
require-await rule doesn't accept for-await-of loop

### ISSUE THREAD

**Issue by Jamesernator** [eslint/eslint#8300]
* **ESLint Version:** 3.18.0
* **Node Version:** 7.7.3
* **npm Version:** 4.1.2

**What parser (default, Babel-ESLint, etc.) are you using?** babel-eslint

Configuration:

```
env:
    node: true
    browser: true
    commonjs: true
    es6: true
parser: babel-eslint
parserOptions:
    sourceType: module
    ecmaVersion: 2017
rules:
    # Possible errors
    no-await-in-loop:
        - off
    no-cond-assign:
        - error
    no-console:
        - off
    no-constant-condition:
        - warn
    no-control-regex:
        - error
    no-debugger:
        - error
    no-dupe-keys:
        - error
    no-duplicate-case:
        - error
    no-empty-character-class:
        - error
    no-empty:
        - warn
    no-ex-assign:
        - error
    no-extra-boolean-cast:
        - error
    no-extra-parens:
        - error
    no-extra-semi:
        - error
    no-func-assign:
        - error
    no-inner-declarations:
        - off
    no-invalid-regexp:
        - error
    no-irregular-whitespace:
        - error
    no-obj-calls:
        - error
    no-prototype-builtins:
        - error
    no-regex-spaces:
        - error
    no-sparse-arrays:
        - off
    no-template-curly-in-string:
        - warn
    no-unexpected-multiline:
        - warn
    no-unreachable:
        - error
    no-unsafe-finally:
        - error
    no-unsafe-negation:
        - error
    use-isnan:
        - error
    valid-typeof:
        - error

    # Best Practices
    accessor-pairs:
        - off
    array-callback-return:
        - error
    block-scoped-var:
        - error
    class-methods-use-this:
        - warn
    complexity:
        - warn
        - max: 40
    consistent-return:
        - warn
    curly:
        - error
        - all
    default-case:
        - off
    dot-location:
        - error
        - property
    dot-notation:
        - error
    eqeqeq:
        - error
        - "smart"
    guard-for-in:
        - error
    no-alert:
        - off
    no-caller:
        - error
    no-case-declarations:
        - error
    no-div-regex:
        - off
    no-else-return:
        - off
    no-empty-function:
        - warn
    no-empty-pattern:
        - error
    no-eq-null:
        - off
    no-eval:
        - off
    no-extend-native:
        - error
    no-extra-bind:
        - error
    no-extra-label:
        - error
    no-fallthrough:
        - error
    no-floating-decimal:
        - error
    no-global-assign:
        - error
    no-implicit-coercion:
        - warn
    no-implicit-globals:
        - error
    no-implied-eval:
        - error
    no-invalid-this:
        - warn
    no-iterator:
        - error
    no-labels:
        - warn
    no-lone-blocks:
        - error
    no-loop-func:
        - error
    no-magic-numbers:
        - error
        - ignore: [-1, 0, 1, 2]
    no-multi-spaces:
        - error
    no-multi-str:
        - error
    no-new-func:
        - warn
    no-new-wrappers:
        - error
    no-new:
        - error
    no-octal-escape:
        - error
    no-octal:
        - error
    no-param-reassign:
        - warn
        - props: true
    no-proto:
        - error
    no-redeclare:
        - error
        - builtinGlobals: true
    no-restricted-properties:
        - off
    no-return-assign:
        - error
    no-return-await:
        - error
    no-script-url:
        - error
    no-self-assign:
        - error
    no-self-compare:
        - error
    no-sequences:
        - error
    no-throw-literal:
        - error
    no-unmodified-loop-condition:
        - error
    no-unused-expressions:
        - error
    no-unused-labels:
        - error
    no-useless-call:
        - error
    no-useless-concat:
        - error
    no-useless-escape:
        - error
    no-useless-return:
        - error
    no-void:
        - warn
    no-warning-comments:
        - error
    no-with:
        - error
    prefer-promise-reject-errors:
        - warn
    radix:
        - error
        - always
    require-await:
        - error
    vars-on-top:
        - off
    wrap-iife:
        - error
        - any
    yoda:
        - error
        - never
        - exceptRange: true
          onlyEquality: true

    # Strict mode
    strict:
        - off

    # Variables
    init-declarations:
        - off
    no-catch-shadow:
        - error
    no-delete-var:
        - error
    no-label-var:
        - error
    no-restricted-globals:
        - off
    no-shadow-restricted-names:
        - error
    no-shadow:
        - warn
    no-undef-init:
        - error
    no-undef:
        - error
        - typeof: true
    no-undefined:
        - off
    no-unused-vars:
        - error
        - argsIgnorePattern: "^_"
          varsIgnorePattern: "^_"
          caughtErrorsIgnorePattern: "^_"
    no-use-before-define:
        - warn

    # CommonJS
    callback-return:
        - error
    global-require:
        - error
    handle-callback-err:
        - error
    no-mixed-requires:
        - error
    no-new-require:
        - error
    no-path-concat:
        - error
    no-process-env:
        - error
    no-process-exit:
        - error
    no-restricted-modules:
        - off
    no-sync:
        - off

    # Style
    array-bracket-spacing:
        - off
    block-spacing:
        - error
    brace-style:
        - error
    camelcase:
        - error
    capitalized-comments:
        - off
    comma-dangle:
        - off
    comma-spacing:
        - off
    comma-style:
        - off
    computed-property-spacing:
        - error
    consistent-this:
        - error
    eol-last:
        - error
    func-call-spacing:
        - off
    func-name-matching:
        - warn
    func-names:
        - off
    func-style:
        - off
    id-blacklist:
        - off
    id-length:
        - off
    id-match:
        - off
    indent:
        - off
    jsx-quotes:
        - off
    key-spacing:
        - error
    keyword-spacing:
        - error
    line-comment-position:
        - off
    linebreak-style:
        - off
    lines-around-comment:
        - off
    lines-around-directive:
        - off
    max-depth:
        - error
        - max: 4
    max-len:
        - error
        - code: 120
          tabWidth: 4
          ignoreUrls: true
          ignoreComments: true
          ignoreStrings: true
          ignoreTemplateLiterals: true
          ignoreRegExpLiterals: true
          ignoreTrailingComments: true
    max-lines:
        - off
    max-nested-callbacks:
        - error
        - max: 3
    max-params:
        - error
        - max: 5
    max-statements-per-line:
        - off
    max-statements:
        - error
        - max: 30
    multiline-ternary:
        - off
    new-cap:
        - error
    new-parens:
        - error
    newline-after-var:
        - off
    newline-before-return:
        - off
    newline-per-chained-call:
        - error
        - ignoreChainWithDepth: 2
    no-array-constructor:
        - error
    no-bitwise:
        - error
    no-continue:
        - off
    no-inline-comments:
        - off
    no-lonely-if:
        - error
    no-mixed-operators:
        - off
    no-mixed-spaces-and-tabs:
        - error
    no-multi-assign:
        - error
    no-multiple-empty-lines:
        - error
        - max: 10
    no-negated-condition:
        - error
    no-nested-ternary:
        - error
    no-new-object:
        - error
    no-plusplus:
        - error
        - allowForLoopAfterthoughts: true
    no-restricted-syntax:
        - off
    no-tabs:
        - warn
    no-ternary:
        - off
    no-trailing-spaces:
        - warn
    no-underscore-dangle:
        - off
    no-unneeded-ternary:
        - error
    no-whitespace-before-property:
        - error
    object-curly-newline:
        - off
    object-curly-spacing:
        - error
        - always
    object-property-newline:
        - error
        - allowMultiplePropertiesPerLine: true
    one-var-declaration-per-line:
        - off
    one-var:
        - off
    operator-assignment:
        - error
    operator-linebreak:
        - error
        - before
        - overrides:
            "=": ignore
            "?": ignore
            ":": ignore
    padded-blocks:
        - off
    quote-props:
        - off
    quotes:
        - off
    require-jsdoc:
        - off
    semi-spacing:
        - error
        - before: true
          after: true
    semi:
        - off
    sort-keys:
        - off
    sort-vars:
        - off
    space-before-blocks:
        - error
    space-before-function-paren:
        - off
    space-in-parens:
        - off
    space-infix-ops:
        - off
    space-unary-ops:
        - error
        - words: true
          nonwords: false
    spaced-comment:
        - off
    template-tag-spacing:
        - error
    unicode-bom:
        - off
    wrap-regex:
        - off

    # ES6+
    arrow-body-style:
        - off
    arrow-parens:
        - error
        - as-needed
    arrow-spacing:
        - error
    constructor-super:
        - error
    generator-star-spacing:
        - warn
        - before: false
          after: false
    no-class-assign:
        - error
    no-confusing-arrow:
        - error
        - allowParens: true
    no-const-assign:
        - error
    no-dupe-class-members:
        - error
    no-duplicate-imports:
        - error
    no-new-symbol:
        - error
    no-restricted-imports:
        - off
    no-this-before-super:
        - error
    no-useless-computed-key:
        - error
    no-useless-constructor:
        - error
    no-useless-rename:
        - error
    no-var:
        - warn
    object-shorthand:
        - error
    prefer-arrow-callback:
        - error
        - allowNamedFunctions: true
    prefer-const:
        - error
    prefer-destructuring:
        - off
    prefer-numeric-literals:
        - warn
    prefer-rest-params:
        - warn
    prefer-spread:
        - error
    prefer-template:
        - warn
    require-yield:
        - warn
    rest-spread-spacing:
        - error
    sort-imports:
        - off
    symbol-description:
        - error
    template-curly-spacing:
        - off
    yield-star-spacing:
        - warn
        - before: false
          after: true
```

**What did you do? Please include the actual source code causing the issue.**

I was just testing out for-await loops but this simple function doesn't pass the rule require-await

```js
const ticks = async function*(time) {
    for (let i=0 ; ; i++) {
        yield i
        await new Promise(resolve => setTimeout(resolve, time))
    }
}

async function main() {
    const delay = 1000
    for await (const tick of ticks(delay)) {
        console.log(tick)
    }
}

main()
```

**What did you expect to happen?**

I expected that `for-await-of` loop would be sufficient for the rule `require-await` to pass.

**What actually happened? Please include the actual, raw output from ESLint.**

```
/home/jamesernator/Projects/code-playground/for-await.js
  10:1  error  Async function 'main' has no 'await' expression  require-await

✖ 1 problem (1 error, 0 warnings)
```

**Comment by not-an-aardvark** [comment]
Thanks for the report.

for-await-of loops are a [stage 3 proposal](https://github.com/tc39/proposal-async-iteration), but eslint core rules don't support experimental features until they reach stage 4 and officially become part of the spec. For now, I would recommend creating an issue on [eslint-plugin-babel](https://github.com/babel/eslint-plugin-babel) to create a custom rule that works with this experimental syntax.

**Comment by alberto** [comment]
Closing this as it doesn't appear that there is enough interest, either from the team or from the community, to implement this at this time.
